lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,650 @@
+                message.extend_from_slice(chunk);
+            }
+            Some(message.freeze())
+        } else {
+            None
+        }
+    }
+}
+
+/// A codec for handling GELF messages that may be chunked. The implementation is based on [Graylog's GELF documentation](https://go2docs.graylog.org/5-0/getting_in_log_data/gelf.html#GELFviaUDP)
+/// and [Graylog's go-gelf library](https://github.com/Graylog2/go-gelf/blob/v1/gelf/reader.go).
+#[derive(Debug, Clone)]
+pub struct ChunkedGelfDecoder {
+    // We have to use this decoder to read all the bytes from the buffer first and don't let tokio
+    // read it buffered, as tokio FramedRead will not always call the decode method with the
+    // whole message. (see https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#26).
+    // This limitation is due to the fact that the GELF format does not specify the length of the
+    // message, so we have to read all the bytes from the message (datagram)
+    bytes_decoder: BytesDecoder,
+    state: Arc<Mutex<HashMap<u64, MessageState>>>,
+    timeout: Duration,
+    pending_messages_limit: usize,
+}
+
+impl ChunkedGelfDecoder {
+    /// Creates a new `ChunkedGelfDecoder`.
+    pub fn new(timeout_millis: u64, pending_messages_limit: usize) -> Self {
+        Self {
+            bytes_decoder: BytesDecoder::new(),
+            state: Arc::new(Mutex::new(HashMap::new())),
+            timeout: Duration::from_millis(timeout_millis),
+            pending_messages_limit,
+        }
+    }
+
+    /// Decode a GELF chunk
+    pub fn decode_chunk(&mut self, mut chunk: Bytes) -> Result<Option<Bytes>, BoxedFramingError> {
+        // Encoding scheme:
+        //
+        // +------------+-----------------+--------------+----------------------+
+        // | Message id | Sequence number | Total chunks |    Chunk payload     |
+        // +------------+-----------------+--------------+----------------------+
+        // | 64 bits    | 8 bits          | 8 bits       | remaining bits       |
+        // +------------+-----------------+--------------+----------------------+
+        //
+        // As this codec is oriented for UDP, the chunks (datagrams) are not guaranteed to be received in order,
+        // nor to be received at all. So, we have to store the chunks in a buffer (state field) until we receive
+        // all the chunks of a message. When we receive all the chunks of a message, we can concatenate them
+        // and return the complete payload.
+
+        // We need 10 bytes to read the message id, sequence number and total chunks
+        if chunk.remaining() < 10 {
+            let src_display = format!("{chunk:?}");
+            warn!(message = "Received malformed chunk headers (message ID, sequence number and total chunks) with less than 10 bytes. Ignoring it.",
+                src = src_display,
+                remaining = chunk.remaining(),
+                internal_log_rate_limit = true
```

Comment:
Added the `internal_log_rate_limit = true` in every log here, I don't know if its okay